File: web/app/types/scene.ts
// SceneListItem is a lightweight scene representation for list/grid displays.
// Used by: homepage sections, search results, actor/studio scenes, explorer, history
export interface SceneListItem {
    id: number;
    title: string;
    duration: number;
    size: number;
    thumbnail_path: string;
    preview_video_path: string;
    processing_status: string;
    is_corrupted: boolean;
    created_at: string;
    updated_at: string;
    // Optional fields included via card_fields
    view_count?: number;
    width?: number;
    height?: number;
    frame_rate?: number;
    description?: string;
    studio?: string;
    tags?: string[];
    actors?: string[];
}

// Scene is the full scene representation with all metadata.
// Used by: scene player page, scene details editor
export interface Scene extends SceneListItem {
    original_filename: string;
    stored_path: string;
    storage_path_id?: number;
    view_count: number;
    width?: number;
    height?: number;
    sprite_sheet_path?: string;
    vtt_path?: string;
    sprite_sheet_count?: number;
    thumbnail_width?: number;
    thumbnail_height?: number;
    processing_error?: string;
    file_created_at?: string;
    description?: string;
    studio?: string;
    tags?: string[];
    actors?: string[];
    cover_image_path?: string;
    file_hash?: string;
    frame_rate?: number;
    bit_rate?: number;
    video_codec?: string;
    audio_codec?: string;
    release_date?: string;
    porndb_scene_id?: string;
    origin?: string;
    type?: string;
    // Set on shorts cut from another scene
    source_scene_id?: number | null;
    source_start?: number | null;
    source_end?: number | null;
}

export interface SceneListResponse {
    data: SceneListItem[];
    total: number;
    page: number;
    limit: number;
    ratings?: Record<string, number>;
    likes?: Record<string, boolean>;
    jizz_counts?: Record<string, number>;
    seed?: number;
}

export interface SceneSearchParams {
    q?: string;
    tags?: string;
    actors?: string;
    studio?: string;
    min_duration?: number;
    max_duration?: number;
    min_date?: string;
    max_date?: string;
    resolution?: string;
    sort?: string;
    page?: number;
    limit?: number;
    liked?: boolean;
    min_rating?: number;
    max_rating?: number;
    min_jizz_count?: number;
    max_jizz_count?: number;
    origin?: string;
    type?: string;
}

export interface TagOption {
    id: number;
    name: string;
    color: string;
    scene_count: number;
}

export interface MarkerLabelOption {
    label: string;
    count: number;
}

export interface SceneFilterOptions {
    studios: string[];
    actors: string[];
    tags: TagOption[];
    marker_labels: MarkerLabelOption[];
    origins: string[];
    types: string[];
}

File: internal/api/v1/request/scene.go
package request

type UpdateSceneDetailsRequest struct {
	Title       string  `json:"title"`
	Description string  `json:"description"`
	ReleaseDate *string `json:"release_date,omitempty"`
}

type SetRatingRequest struct {
	Rating float64 `json:"rating" binding:"required,min=0.5,max=5"`
}

type SearchScenesRequest struct {
	Query        string  `form:"q"`
	Tags         string  `form:"tags"`
	Actors       string  `form:"actors"`
	Studio       string  `form:"studio"`
	MinDuration  int     `form:"min_duration"`
	MaxDuration  int     `form:"max_duration"`
	MinDate      string  `form:"min_date"`
	MaxDate      string  `form:"max_date"`
	Resolution   string  `form:"resolution"`
	Sort         string  `form:"sort"`
	Page         int     `form:"page"`
	Limit        int     `form:"limit"`
	Liked        *bool   `form:"liked"`
	MinRating    float64 `form:"min_rating"`
	MaxRating    float64 `form:"max_rating"`
	MinJizzCount int     `form:"min_jizz_count"`
	MaxJizzCount int     `form:"max_jizz_count"`
	MatchType    string  `form:"match_type"`
	MarkerLabels string  `form:"marker_labels"` // Comma-separated list of marker labels
	Seed         int64   `form:"seed"`          // Random shuffle seed (0 = auto-generate)
	Shorts       string  `form:"shorts"`        // all (default), only, hide, only_clips or hide_clips
}

type ApplySceneMetadataRequest struct {
	Title         *string  `json:"title,omitempty"`
	Description   *string  `json:"description,omitempty"`
	Studio        *string  `json:"studio,omitempty"`
	ThumbnailURL  *string  `json:"thumbnail_url,omitempty"`
	ActorIDs      []uint   `json:"actor_ids,omitempty"`
	TagNames      []string `json:"tag_names,omitempty"`
	ReleaseDate   *string  `json:"release_date,omitempty"`
	PornDBSceneID *string  `json:"porndb_scene_id,omitempty"`
}

type DeleteSceneRequest struct {
	Permanent bool `json:"permanent"`
}
